File: 2494-sum-of-prefix-scores-of-strings/2494-sum-of-prefix-scores-of-strings.cpp
#include <vector>
#include <string>
#include <unordered_map>

class Solution {
public:
    std::vector<int> sumPrefixScoresOne(std::vector<std::string>& words) {
        // Trie node definition
        class TrieNode {
        public:
            int count;
            std::unordered_map<char, TrieNode*> children;
            TrieNode() : count(0) {}
        };
        
        // Root of the Trie
        TrieNode* root = new TrieNode();
        
        // Build the Trie and count prefix occurrences
        for (const std::string& word : words) {
            TrieNode* curr = root;
            for (char c : word) {
                if (curr->children.find(c) == curr->children.end()) {
                    curr->children[c] = new TrieNode();
                }
                curr = curr->children[c];
                curr->count += 1;
            }
        }
        
        // Calculate prefix scores for each word
        std::vector<int> prefixCounts;
        for (const std::string& word : words) {
            int prefixCount = 0;
            TrieNode* curr = root;
            for (char c : word) {
                curr = curr->children[c];
                prefixCount += curr->count;
            }
            prefixCounts.push_back(prefixCount);
        }
        
        // Cleanup (optional)
        // You would typically need to clean up the allocated Trie nodes here.
        // For now, we'll skip the cleanup for simplicity.

        return prefixCounts;
    }

    	vector<int> sumPrefixScores(vector<string>& words) {
		int n = words.size();
		vector<pair<string, int>> words2;
		for (int i = 0; i < n; ++i) {
			words2.push_back(make_pair(words[i], i));
		}
		sort(words2.begin(), words2.end());

		vector<int> perm(n);
		for (int i = 0; i < n; ++i) {
			perm[i] = words2[i].second;
		}
		vector<int> commonPrefix(n);

		for (int i = 1; i < n; ++i) {
			string const& w1 = words2[i - 1].first;
			string const& w2 = words2[i].first;
			int l = min(w1.size(), w2.size());
			int p = 0;
			while (p < l && w1[p] == w2[p]) {
				++p;
			}
			commonPrefix[i] = p;
		}

		vector<int> ret(n);
		for (int i = 0; i < n; ++i) {
			int prefix = words2[i].first.size();
			ret[perm[i]] += prefix;
			for (int j = i + 1; j < n; ++j) {
				prefix = min(prefix, commonPrefix[j]);
				if (prefix == 0) {
					break;
				}
				ret[perm[i]] += prefix;
				ret[perm[j]] += prefix;
			}
		}
		return ret;
	}
};

File: 2494-sum-of-prefix-scores-of-strings/2494-sum-of-prefix-scores-of-strings.py
class TrieNode:
    def __init__(self):
        self.count = 0
        self.child = [None] * 26


class Solution:
    @cache
    def calcPrefix(self, word: str, words: tuple) -> int:
        count=0
        for w in words:
            if(w.startswith(word)):
                count+=1
        return count

    def sumPrefixScoresSlow(self, words: List[str]) -> List[int]:
        ret=[]
        tuple_world = tuple(words)
        for word in words:
            r=0
            tmp=""
            for c in word:
                tmp+=c
                rr = self.calcPrefix(tmp, tuple_world)
                r+= rr
                if rr == 1:
                    r+=len(word)-len(tmp)
                    break
            ret.append(r)
        return ret

    def sumPrefixScores(self, words: List[str]) -> List[int]:
        def insert_key(root:TrieNode, key:str):
            curr = root
            for c in key:
                index = ord(c) - ord('a')
                if curr.child[index] is None:
                    new_node = TrieNode()
                    new_node.count = 1
                    curr.child[index] = new_node
                else:
                    curr.child[index].count += 1
                curr = curr.child[index]
        @cache
        def find_score(root:TrieNode, key:str) -> int:
            curr = root
            ret = 0
            for c in key:
                index = ord(c) - ord('a')
                ret+=curr.count
                if curr.child[index] is None:
                    return ret
                curr = curr.child[index]
            ret += curr.count
            return ret

        ret=[]
        root = TrieNode()
        for w in words:
            insert_key(root, w)
        print(root.count)
        for w in words:
            ret.append(find_score(root,w))
        return ret


        